pubsub/gcp/pubsub/pubsub.go

```diff
@@ -0,0 +1,149 @@
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	gcppubsub "cloud.google.com/go/pubsub"
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/dapr/pkg/logger"
+	"google.golang.org/api/option"
+)
+
+const (
+	errorMessagePrefix = "gcp pubsub error:"
+)
+
+// GCPPubSub type
+type GCPPubSub struct {
+	client   *gcppubsub.Client
+	metadata *metadata
+	logger   logger.Logger
+}
+
+// NewGCPPubSub returns a new GCPPubSub instance
+func NewGCPPubSub(logger logger.Logger) pubsub.PubSub {
+	return &GCPPubSub{logger: logger}
+}
+
+// Init parses metadata and creates a new Pub Sub client
+func (g *GCPPubSub) Init(meta pubsub.Metadata) error {
+	b, err := g.parseMetadata(meta)
+	if err != nil {
+		return err
+	}
+
+	var pubsubMeta metadata
+	err = json.Unmarshal(b, &pubsubMeta)
+	if err != nil {
+		return err
+	}
+	clientOptions := option.WithCredentialsJSON(b)
+	ctx := context.Background()
+	pubsubClient, err := gcppubsub.NewClient(ctx, pubsubMeta.ProjectID, clientOptions)
+	if err != nil {
+		return fmt.Errorf("%s error creating pubsub client: %s", errorMessagePrefix, err)
+	}
+
+	if pubsubMeta.Subscription == "" {
```

Comment:
Dapr will populate the `pubsub.Metadata` with a `consumerID` key. That should be used for the Subscription.

You can see an example here: https://github.com/dapr/components-contrib/blob/cf077cca65694b53088dc726603d625f28135faa/pubsub/azure/servicebus/servicebus.go#L184